rfcs/2021-08-29-8381-vrl-iteration-support.md

```diff
@@ -90,28 +102,380 @@ One gap in the language right now is the possibility to _dynamically remap
+```
+
+###### details
+
+The function works similarly to `map_keys`, except that it maps the values
+instead of keys, and thus can also be used to map values within arrays.
+
+Recursion behaves similarly to `map_keys` as well.
+
+##### `for_each`
+
+Iterate over objects or arrays, without mutating any data.
+
+###### function signature
+
+```coffee
+for_each(value: object|array, recursive: bool) -> |string OR integer, any| { any }
+```
+
+###### details
+
+This can be considered a ”trap door” iteration function that allows you to
+tackle any use-case not solved by any of the existing (or future) specialized
+iteration functions.
+
+The drawback of such a function is that it potentially requires more manual
+”set-up” code to get the end-result (e.g. initializing empty collections to
+populate during a `for_each` run, for example).
+
+As the name implies, this function does not mutate the given collection, and
+instead always returns `null`. It can be used to mutate data external to the
+closure, while iterating over the collection. In a sense, it’s the most
+general-purpose iteration function that allows you to manually write mapping,
+reducing, filtering or counting logic.
+
+##### why no `map` function exists
+
+Some might note that there’s no `map` function, only specialized `map_keys` and
+`map_values`.
+
+The reason for this omission is that `map` becomes complicated when dealing
+with recursion, and the closure signature differs when dealing with an object or
+array, requiring us to know at compile-time the exact type of the iteration
+target.
+
+In the end, all current requested use-cases by operators could be solved by one
+of the three proposed iteration functions, allowing us to skip the additional
+work of figuring out how `map` would work exactly, until there’s an actual need
+for such a function (if ever).
+
+### Use Cases
+
+What follows is a list of reported use-cases, and a valid program that uses
+iteration to solve that use-case. Note that there are multiple ways to solve
+individual use-cases, this list shows one available solution per use-case.
+
+1. [nullify empty strings](https://discord.com/channels/742820443487993987/746070591097798688/950750550583050321)
+
+   ```coffee
+   . = map_values(., recursive: true) -> |value| { if value == "" { null } else { value } }
+   ```
+
+2. [converting a single metric into multiple metrics](https://discord.com/channels/742820443487993987/746070591097798688/946148752073326623)
+
+   ```coffee
+   . = { "id": "booster", "timestamp": 123456, "data": { "acceleration": 10, "velocity": 20 } }
+
+   data = del(.data)
+   metrics = []
+   for_each(data) -> |key, value| {
+     .value = value
+     metrics = push(metrics, .)
+   }
+   ```
+
+3. [de-dot keys for Elasticsearch](https://discord.com/channels/742820443487993987/764187584452493323/940359777958121504)
+
+   ```coffee
+   . = map_keys(., recursive: true) -> |key| { replace(key, ".", "_") }
+   ```
```

Comment:
Again, this is great as a iteration use case but, ultimately, I'd like to see us provide special purpose functions for these things (i.e., `replace_keys`).

Reply:
Agreed, the opportunity for functions like these is mentioned in the RFC. We'll get to that once RFC lands.